clustered/server/src/main/java/org/ehcache/clustered/server/store/ClusterTierActiveEntity.java

```diff
@@ -586,38 +603,166 @@ public void synchronizeKeyToPassive(PassiveSynchronizationChannel<EhcacheEntityM
+   * the chains and number of keys in the map.
+   * <p>
+   * NOTE: if expected gets dips below 32, keep at 32 as it indicates a large segment with possibly smaller number of keys
+   * which means the next iteration may show more keys in the map.
+   *
+   * @param remainingSize is the number of keys left in the segment yet to be send
+   * @param expectedGets is the max expected number of keys that could be put in the map before the map gets full
+   * @return required capacity for the map.
+   */
+  private int computeMapCapacity(int remainingSize, int expectedGets) {
+    if (remainingSize < 16) {
+      return 16;
+    } else if (expectedGets < 32) {
+      return 32;
+    } else if (remainingSize < expectedGets) {
+      return (int) ((float) remainingSize / 0.75f + 1.0f);
+    } else {
+      return (int) ((float) expectedGets / 0.75f + 1.0f);
+    }
+  }
+
+  /**
+   * Adjust {@link this#dataMapInitialCapacity} based on what we learned about the cache during iteration of segment 0.
+   *
+   * NOTE: The required capacity calculation and the initial capacity adjustment assumes some sort of symmetry across
+   * multiple segments, but it is possible that in a given segment, some keys has chain with LARGE data and some keys
+   * has small chain with smaller data sizes. But on a larger sweep that should even out. Even if it does not even out,
+   * this should perform better as the initial size is reset back to a minimum of 32 and not 16 when a cache is large
+   * and when a cache is very small it starts with initial size of 16 as the {@link this#computeInitialMapCapacity(int)}
+   * will take the total number of keys in the segment into account.
+   *
+   * @param actualKeyGets the actual number of keys we got when the map got full
+   */
+  private void adjustInitialCapacity(int actualKeyGets) {
+    // even when there are larger data chains with less keys..let us keep the lower bound at 32.
+    dataMapInitialCapacity =  (actualKeyGets < 32) ? 32 : (int) ((float) actualKeyGets / 0.75f + 1.0f);
+  }
+
+  /**
+   * Starts with an initial size of configured {@link this#dataGetsThreshold} or adjusted initial size, unless the
+   * segment of the cache is smaller than the initial expected size.
+   *
+   * @param totalKeys is the total number of keys in this segment
+   */
+  private int computeInitialMapCapacity(int totalKeys) {
+    if (dataMapInitialCapacity == null) {
+      dataMapInitialCapacity = (int) ((float) dataGetsThreshold / 0.75f + 1.0f);
+    }
+    if (totalKeys < 16) {
+      return 16;
+    } else if (totalKeys < dataMapInitialCapacity) {
+      return (int) ((float) totalKeys / 0.75f + 1.0f);
+    } else {
+      return dataMapInitialCapacity;
+    }
+  }
+
+  /**
+   * Executes message sending asynchronously to preparation of message.
+   */
+  @FunctionalInterface
+  private interface DataSyncMessageHandler {
```

Comment:
Can't we just use a Callable<Boolean>?

Reply:
Yes..Will do that.

Reply:
Up to you... this is a minor nit for me.